doc/inlining.md

```diff
@@ -52,74 +78,122 @@ to:
   version opt
-    1   var x = 1
-    2a  var z_1 = x
-    2b  osr sp2 [] (foo, base, sp2) [var z = z_1]   # wrong
-    2c  var y = z_1
-    2d  drop x_1
-    3   osr sp1 [] (main, base, sp1) [var x = x, var y = y]
-    4   print y
-  ...
-
-function foo(var z)
-  version base
-    1   osr sp2 [] (foo, base, sp2) [var z = z]
-    2   return z
+        var x = 1
+        # prologue
+        var z' = z
+        var y = nil
+    i1: assume [...] or (bar, base, checkpoint) varmap, cont-frames # wrong
+    i2:
+        y <- z'
+        drop z'
+        goto l
+    l:
+        print y
 ```
 
-For the recreation of the stack to work the precise stacklayout of the caller needs to be known after the inlinee returns (ie. there must be an osr right after the call).
-The combined osr is supposed to push the continuation for the call to foo.
+This is incorrect, because we would jump into (bar, base, checkpoint)
+with one less frame on the stack than previously. Before inlining, the
+bailout would end up with `main()`, then `foo()`, then the frames of
+`cont-frames` on the stack. After this broken inlining you would only
+get `main(), cont-frames`. We need to recreate a frame for `foo()`.
 
-```
-    2b  osr sp2 [] (foo, base, sp2) [var z = z_1],
-                   (main, base, sp1) [var y = $, var x = x]
-```
+A continuation frame description contains:
 
-The special $ register specifies the target binding for the continuation.
+- the return variable `y`
 
-Remember the following defs from the semantic:
+- caller version (`main/base`)
 
-    C ::= (a, E, I, l)                 continuation:  result accumulator, environment, instructions, label
-    M ::= (P, C*) : (I, T, H, E, l)    machine state: program, continuations : instructions, trace, heap, environment, pc
+- the return label label `l`, which will be turned into a return pc
+  when building the frame at runtime
 
-Before the safepoint `sp2` our machine state is:
+- the environment at the call point (or, really, the environment
+  needed after returning)
 
-    (P, ())                          :  (main::opt, T, H, {x=1, z_1=1}, 2b)
+To know what the environment at the call-point is, we could decide to
+only inline function calls that are followed by a checkpoint; the
+call-point environment would be exactly the checkpoint
+environment. But in fact we don't need this restriction; during
+inlining, the current dynamic environment for each instruction iN can
+always be split into two components:
 
-If the osr fires the next state would be:
+- the environment of the caller, E{caller}
+- the environment of the inlined code, E{inlined}N
 
-    (P, (y, {x=1}, main::base, sp1)) :  (foo::base, T, H, {z=1},         1)
+At any label `iN`, the current environment is the disjoint union of
+`E{caller}` and `E{inlined}N`. We know that `E{caller}` is unchanged during
+the inlined function execution, given that those instructions come
+from the callee where it is not in scope, and thus cannot modify it.
```

Comment:
I think this is orthogonal: this section is talking carefully about how the scopes evolve on `osr` instructions inlined. Your remark concerns the fact that return instructions (explicit or implicit) are translated in a scope-preserving way, but it is more about the correctness of inlining in general than osr-specific.

I added a remark about that in the comments of the warmup example.